File: src/SustcManager.java
import cs307.project2.interfaces.*;

import java.sql.*;

public class SustcManager implements ISustcManager {
    PreparedStatement loginStatement;
    PreparedStatement itemInfoStatement;
    PreparedStatement shipInfoStatement;
    PreparedStatement itemOfShipStatement;
    PreparedStatement containerInfoStatement;
    PreparedStatement itemOfContainerStatement;
    PreparedStatement staffInfoStatement;

    @Override
    public int getCompanyCount(LogInfo logInfo) {
        if (!login(logInfo)) {
            return -1;
        }
        try {
            Statement statement = getConnection().createStatement();
            ResultSet result = statement.executeQuery("SELECT count(*) as company_count FROM company");
            result.next();
            return result.getInt("company_count");

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public int getCityCount(LogInfo logInfo) {
        if (!login(logInfo)) return -1;
        try {
            Statement statement = getConnection().createStatement();
            ResultSet result = statement.executeQuery("SELECT count(*) as city_count FROM (select name from city union select name from port_city)");
            result.next();
            return result.getInt("city_count");
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public int getCourierCount(LogInfo logInfo) {
        if (!login(logInfo)) {
            return -1;
        }
        try {
            Statement statement = getConnection().createStatement();
            ResultSet result = statement.executeQuery("SELECT count(*) as courier_count FROM courier");
            result.next();
            return result.getInt("courier_count");

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public int getShipCount(LogInfo logInfo) {
        if (!login(logInfo)) {
            return -1;
        }
        try {
            Statement statement = getConnection().createStatement();
            ResultSet result = statement.executeQuery("SELECT count(*) as ship_count FROM ship");
            result.next();
            return result.getInt("ship_count");

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public ItemInfo getItemInfo(LogInfo logInfo, String s) {
        if (!login(logInfo)) {
            return null;
        }
        try {
            if (itemInfoStatement == null) {
                itemInfoStatement = getConnection().prepareStatement("SELECT * FROM item where item.name = ?");
            }
            itemInfoStatement.setString(1, s);
            ResultSet queryResult = itemInfoStatement.executeQuery();
            if (!queryResult.next()) {
                return null;
            }
            ItemState itemState;
            switch (queryResult.getString("state")) {
                case "Delivering":
                    itemState = ItemState.Delivering;
                    break;
                case "Export Check Fail":
                    itemState = ItemState.ExportCheckFailed;
                    break;
                case "Export Checking":
                    itemState = ItemState.ExportChecking;
                    break;
                case "Finish":
                    itemState = ItemState.Finish;
                    break;
                case "From-Import Transporting":
                    itemState = ItemState.FromImportTransporting;
                    break;
                case "Import Check Fail":
                    itemState = ItemState.ImportCheckFailed;
                    break;
                case "Importing Checking":
                    itemState = ItemState.ImportChecking;
                    break;
                case "Packing to Container":
                    itemState = ItemState.PackingToContainer;
                    break;
                case "Picking-up":
                    itemState = ItemState.PickingUp;
                    break;
                case "Shipping":
                    itemState = ItemState.Shipping;
                    break;
                case "To-Export Transporting":
                    itemState = ItemState.ToExportTransporting;
                    break;
                case "Unpacking from Container":
                    itemState = ItemState.UnpackingFromContainer;
                    break;
                case "Waiting for Shipping":
                    itemState = ItemState.WaitingForShipping;
                    break;
                default:
                    itemState = null;
            }

            ItemInfo.RetrievalDeliveryInfo retrieval = new ItemInfo.RetrievalDeliveryInfo(queryResult.getString("from_city_name"), queryResult.getString("retrieval_courier"));
            ItemInfo.RetrievalDeliveryInfo delivery = new ItemInfo.RetrievalDeliveryInfo(queryResult.getString("to_city_name"), queryResult.getString("delivery_courier"));
            ItemInfo.ImportExportInfo $export, $import;
            /*if (queryResult.getString("export_officer") != null)
                $export = new ItemInfo.ImportExportInfo(queryResult.getString("export_city"), queryResult.getString("export_officer"), queryResult.getDouble("export_tax"));
            else $export = null;

            if (queryResult.getString("import_officer") != null)
                $import = new ItemInfo.ImportExportInfo(queryResult.getString("import_city"), queryResult.getString("import_officer"), queryResult.getDouble("import_tax"));
            else $import = null;*/

            $export = new ItemInfo.ImportExportInfo(queryResult.getString("export_city"), queryResult.getString("export_officer"), queryResult.getDouble("export_tax"));
            $import = new ItemInfo.ImportExportInfo(queryResult.getString("import_city"), queryResult.getString("import_officer"), queryResult.getDouble("import_tax"));

            ItemInfo result = new ItemInfo(queryResult.getString("name"), queryResult.getString("class"), queryResult.getDouble("price"), itemState, retrieval, delivery, $import, $export);
            return result;
        } catch (
                SQLException e) {
            throw new RuntimeException(e);
        }

    }

    @Override
    public ShipInfo getShipInfo(LogInfo logInfo, String s) {
        if (!login(logInfo)) {
            return null;
        }
        try {
            if (shipInfoStatement == null) {
                shipInfoStatement = getConnection().prepareStatement("SELECT * from ship where ship.name = ?");
            }

            shipInfoStatement.setString(1, s);
            ResultSet queryResult = shipInfoStatement.executeQuery();

            if (!queryResult.next()) return null;

            // judge whether the ship is sailing
            boolean isSailing;
            if (itemOfShipStatement == null) {
                itemOfShipStatement = getConnection().prepareStatement("SELECT count(*) FROM item where item.ship_name = ? AND item.state == 'Shipping'");
            }
            itemOfShipStatement.setString(1, s);
            ResultSet items = itemInfoStatement.executeQuery();
            items.next();
            isSailing = items.getInt("count") > 0;

            return new ShipInfo(queryResult.getString("name"), queryResult.getString("company_name"), isSailing);

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public ContainerInfo getContainerInfo(LogInfo logInfo, String s) {
        if (!login(logInfo)) {
            return null;
        }

        try {
            if (containerInfoStatement == null) {
                containerInfoStatement = getConnection().prepareStatement("SELECT * FROM container where container.code = ?");
            }
            containerInfoStatement.setString(1, s);
            ResultSet queryResult = containerInfoStatement.executeQuery();

            if (!queryResult.next()) return null;

            // check whether it is in use
            if (itemOfContainerStatement == null) {
                itemOfContainerStatement = getConnection().prepareStatement("SELECT count(*) from item where item.container_code = ? AND item.state != 'Finish'");
            }
            itemOfContainerStatement.setString(1, s);
            ResultSet itemQueryResult = itemOfContainerStatement.executeQuery();
            itemQueryResult.next();
            boolean inUse = itemQueryResult.getInt("count") > 0;

            // get its type
            ContainerInfo.Type type;
            switch (queryResult.getString("type")) {
                case "Dry Container":
                    type = ContainerInfo.Type.Dry;
                    break;
                case "Flat Rack Container":
                    type = ContainerInfo.Type.FlatRack;
                    break;
                case "ISO Tank Container":
                    type = ContainerInfo.Type.ISOTank;
                    break;
                case "Open Top Container":
                    type = ContainerInfo.Type.OpenTop;
                    break;
                case "Reefer Container":
                    type = ContainerInfo.Type.Reefer;
                    break;
                default:
                    type = null;

            }

            ContainerInfo containerInfo = new ContainerInfo(type, queryResult.getString("code"), inUse);

            return containerInfo;

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public StaffInfo getStaffInfo(LogInfo logInfo, String s) {
        if (!login(logInfo)) {
            return null;
        }
        try {
            if (staffInfoStatement == null) {
                staffInfoStatement = getConnection().prepareStatement("SELECT * FROM courier where courier.name = ? union SELECT * FROM company_manager where company_name.name = ?");
            }
            staffInfoStatement.setString(1, s);
            staffInfoStatement.setString(2, s);

            ResultSet queryResult = staffInfoStatement.executeQuery();
            if (!queryResult.next()) return null;
            boolean isFemale = queryResult.getString("gender").equals("female");

            return new StaffInfo(logInfo, queryResult.getString("company_name"), queryResult.getString("city_name"), isFemale, queryResult.getInt("age"), queryResult.getString("phone_number"));
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private boolean login(LogInfo logInfo) {
        if (logInfo.type() != LogInfo.StaffType.SustcManager) {
            return false;
        }

        String password = Util.autoEncryptPassword(logInfo.password());
        try {
            if (loginStatement == null) {
                loginStatement = getConnection().prepareStatement("SELECT * FROM sustc_manager WHERE  name = ? AND password = ?");
            }

            loginStatement.setString(1, logInfo.name());
            loginStatement.setString(2, password);

            ResultSet result = loginStatement.executeQuery();
            return result.next();

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private Connection getConnection() {
        return ConnectionManager.getSustcManagerConnection();
    }
}

File: CS307_Project_FrontEnd/sustcManager.py
import util as u
import requests
from config import config
from urllib.parse import urljoin
from util import inputByType as iB

role = "sustc manager"


def sustcManager(username, passwd):
    info = ""
    flag = True

    while flag:
        u.clear()
        print()
        print("Hello! %s: %s" % (role, username))
        print("============================================================")
        print("|                       [Operations]                       |")
        print("|      Please Enter Number To Select Your Operation        |")
        print("|----------------------------------------------------------|")
        print("|                   1. Get Company Count                   |")
        print("|                     2. Get City Count                    |")
        print("|                   3. Get Courier Count                   |")
        print("|                     4. Get Ship Count                    |")
        print("|                     5. Get Item Info                     |")
        print("|                     6. Get Ship Info                     |")
        print("|                   7. Get Container Info                  |")
        print("|                     8. Get Staff Info                    |")
        print("|----------------------------------------------------------|")
        print("|          X. Log Out                     Q. Quit          |")
        print("============================================================")
        print(info)
        print(">>> ", end="")
        op = input()
        if op == '1':
            company_count(username, passwd)
        elif op == '2':
            city_count(username, passwd)
        elif op == '3':
            courier_count(username, passwd)
        elif op == '4':
            ship_count(username, passwd)
        elif op == '5':
            itemName = iB("Item Name")
            item_info(username, passwd, itemName)
        elif op == '6':
            shipName = iB("Ship Name")
            ship_info(username, passwd, shipName)
        elif op == '7':
            containerCode = iB("Container Code")
            container_info(username, passwd, containerCode)
        elif op == '8':
            staffName = iB("Staff Name")
            staff_info(username, passwd, staffName)
        elif op == 'X':
            flag = False
        elif op == 'Q':
            u.exiting()
        else:
            info = "<Wrong Selection>"


def company_count(username, passwd):
    result = requests.get(urljoin(config['base'], '/sustc_manager/company_count'), headers={
        'username': username,
        'password': passwd,
        'role': role
    }).json()
    if result != -1:
        print("Company Count: {0}".format(result))
    else:
        print("Failed!")
    input("Press Enter to continue...")


def city_count(username, passwd):
    result = requests.get(urljoin(config['base'], '/sustc_manager/city_count'), headers={
        'username': username,
        'password': passwd,
        'role': role
    }).json()
    if result != -1:
        print("City Count: {0}".format(result))
    else:
        print("Failed!")
    input("Press Enter to continue...")


def courier_count(username, passwd):
    result = requests.get(urljoin(config['base'], '/sustc_manager/courier_count'), headers={
        'username': username,
        'password': passwd,
        'role': role
    }).json()
    if result != -1:
        print("Courier Count: {0}".format(result))
    else:
        print("Failed!")
    input("Press Enter to continue...")


def ship_count(username, passwd):
    result = requests.get(urljoin(config['base'], '/sustc_manager/ship_count'), headers={
        'username': username,
        'password': passwd,
        'role': role
    }).json()
    if result != -1:
        print("Ship Count: {0}".format(result))
    else:
        print("Failed!")
    input("Press Enter to continue...")


def item_info(username, passwd, itemName):
    result = requests.get(urljoin(config['base'], '/sustc_manager/item_info/{}'.format(itemName)), headers={
        'username': username,
        'password': passwd,
        'role': role
    }).json()
    if result is not None:
        print("Name:\t{1}\n"
              "Class:\t{2}\n"
              "Price:\t{3}\n"
              "State:\t{4}\n"
              "Retrieval City:\t{5}\n"
              "Retrieval Courier:\t{6}\n"
              "Delivery City:\t{7}\n"
              "Delivery Courier:\t{8}\n"
              "Export City:\t{9}\n"
              "Export Officer:\t{10}\n"
              "Export Tax:\t{11}\n"
              "Import City:\t{12}\n"
              "Import Officer:\t{13}\n"
              "Import Tax:\t{14}\n".format(
                None,
                result['name'],
                result['$class'],
                result['price'],
                result['state'],
                result['retrieval']['city'],
                result['retrieval']['courier'],
                result['delivery']['city'],
                result['delivery']['courier'],
                result['export']['city'],
                result['export']['officer'],
                result['export']['tax'],
                result['$import']['city'],
                result['$import']['officer'],
                result['$import']['tax']
              )
              )
    else:
        print("Failed!")
    input("Press Enter to continue...")


def ship_info(username, passwd, shipName):
    result = requests.get(urljoin(config['base'], '/sustc_manager/ship_info/{}'.format(shipName)), headers={
        'username': username,
        'password': passwd,
        'role': role
    }).json()
    if result is not None:
        print("Name:\t{0}\n"
              "Owner:\t{1}\n"
              "Sailing:\t{2}\n".format(
                result['name'],
                result['owner'],
                result['sailing']))
    else:
        print("Failed!")
    input("Press Enter to continue...")


def container_info(username, passwd, containerCode):
    result = requests.get(urljoin(config['base'], '/sustc_manager/container_info/{}'.format(containerCode)), headers={
        'username': username,
        'password': passwd,
        'role': role
    }).json()
    if result is not None:
        print("Code:\t{0}\n"
              "Type:\t{1}\n"
              "Using:\t{2}\n".format(result['code'], result['type'], result['using']))
    else:
        print("Failed!")
    input("Press Enter to continue...")


def staff_info(username, passwd, staffName):
    result = requests.get(urljoin(config['base'], '/sustc_manager/staff_info/{}'.format(staffName)), headers={
        'username': username,
        'password': passwd,
        'role': role
    }).json()
    if result is not None:
        print(result)
        print("Username:\t{}\n"
              "Role:\t{}\n"
              "City:\t{}\n"
              "Is Female:\t{}\n"
              "Age:\t{}\n"
              "Phone Number:\t{}\n".format(
                result['basicInfo']['name'],
                result['basicInfo']['type'],
                result['city'],
                result['isFemale'],
                result['age'],
                result['phoneNumber']))
    else:
        print("Failed!")
    input("Press Enter to continue...")
